### PR TITLE
Add test to scan all python packages

### LINKED ISSUE #1612: Develop an automatic test/tool to detect new security flaws in the embedded python or dependencies
Hi team,

We are looking to develop a tool or test that allows us to automate the process of analyzing security flags in the Python embedded with Wazuh and in the installed deps.

The goal is to be able to launch this tool or test manually every X days at first and finally automate this process for each PR in the main repo.

For the development of this tool or test, we can use tools such as [Safety](https://pypi.org/project/safety/) or any other that allows us to automate the process and generate a report to later analyze it and determine if there is any new vulnerability.

The tool should review the requeriments.txt file for possible vulnerabilities, in addition to the current Python version, to determine if there is any type of warning.

Regards

### PR DESCRIPTION
**Pull request by Kondent**
|Related issue|
|---|
|#1612|

## Description
This PR closes #1612.
I made a tool to check vulnerabilities in all python packages used by wazuh specified from a requirements file or directly from pip.

```
⬦⬦⬦ ~/git/wazuh-qa/tests/system/security ⨘ python3 package_vuln_scan.py -h
usage: package_vuln_scan.py [-h] (-r INPUT | -p) -o OUTPUT

optional arguments:
  -h, --help  show this help message and exit
  -r INPUT    specify requirements file path.
  -p          enable pip scan mode.
  -o OUTPUT   specify output file.
```

## Example of use
```
⬦⬦⬦ ~/git/wazuh-qa/tests/system/security ⨘ python3 package_vuln_scan.py -p -o prueba.txt
⬦⬦⬦ ~/git/wazuh-qa/tests/system/security ⨘ cat prueba.txt 
{
    "report_date": "26/07/2021 - 17:37:45",
    "vulnerabilities_found": 1,
    "packages": [
        {
            "package_name": "insecure-package",
            "package_version": "0.1.0",
            "package_affected_version": "<0.2.0",
            "vuln_description": "This is an insecure package with lots of exploitable security vulnerabilities.",
            "safety_id": "25853"
        }
    ]
}
```

### CODE REVIEW

File: tests/system/security/package_vuln_scan.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import argparse
+import subprocess
+import json
+import re
+import os
+from datetime import datetime
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:6]: Optimize imports

File: tests/system/security/package_vuln_scan.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    """
+    Summary:
+        Create temp input file with all packages listed and prepared to be scanned later on.
+    Args:
+        pip_mode ([bool]): enable/disable pip freeze to retrieve package information.
+        input_file_path ([str]): path to the input file (used if pip_mode is disabled).
+    """
+    with open(INPUT_FILE_PATH_TMP, mode='a') as input_file_tmp:
+        python_process = subprocess.run(['python',
+                                        '--version'],
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:71]:
```suggestion
                                         '--version'],
```

File: tests/system/security/package_vuln_scan.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+def run_report(output_file_path, input_file_path):
+    """
+    Summary:
+        Perform vulnerability scan using Safety to check all packages listed.
+    Args:
+        output_file_path ([str]): file path where results will be saved.
+        input_file_path ([str]): file path where packages are listed.
+    """
+    json_report = []
+    json_data = []
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:36]:
```suggestion
```

File: tests/system/security/package_vuln_scan.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+from datetime import datetime
+
+INPUT_FILE_PATH_TMP = 'input_file_tmp.txt'
+
+
+def get_args():
+    """
+    Summary:
+        Command line argument parsing method.
+    """
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:15]: Change comments to numpy format

File: tests/system/security/package_vuln_scan.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def run_report(output_file_path, input_file_path):
+    """
+    Summary:
+        Perform vulnerability scan using Safety to check all packages listed.
+    Args:
+        output_file_path ([str]): file path where results will be saved.
+        input_file_path ([str]): file path where packages are listed.
+    """
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:34]: Change comments to numpy format

File: tests/system/security/package_vuln_scan.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def prepare_input(pip_mode, input_file_path):
+    """
+    Summary:
+        Create temp input file with all packages listed and prepared to be scanned later on.
+    Args:
+        pip_mode ([bool]): enable/disable pip freeze to retrieve package information.
+        input_file_path ([str]): path to the input file (used if pip_mode is disabled).
+    """
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:68]: Change comments to numpy format

File: tests/system/security/package_vuln_scan.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                        line = line.split(';')[0] + '\n'
+                    tmp = tmp + line
+                input_file_tmp.write(tmp)
+
+
+def cleanup():
+    """
+    Summary:
+        Clean temporal files.
+    """
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:98]: Change comments to numpy format

File: tests/system/security/package_vuln_scan.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    for package_information in json.loads(safety_process.stdout):
+        json_report.append({
+            'package_name': package_information[0],
+            'package_version': package_information[2],
+            'package_affected_version': package_information[1],
+            'vuln_description': package_information[3],
+            'safety_id': package_information[4]
+        })
+    json_data = {
+        'report_date': datetime.now().strftime('%d/%m/%Y - %H:%M:%S'),
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:53]: For dates, use ISO 8601 format

File: tests/system/security/package_vuln_scan.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        python = python_process.stdout.replace(' ', '==')
+        input_file_tmp.write(python)
+        if pip_mode:
+            subprocess.run(['pip', 'freeze'], stdout=input_file_tmp, universal_newlines=True)
+        else:
+            with open(input_file_path, mode='r') as input_file:
+                lines = input_file.readlines()
+                tmp = ''
+                for line in lines:
+                    line = re.sub('[<|>|~]', '=', line)
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:83]:
```suggestion
                    line = re.sub('[<|>~]', '=', line)
```

File: tests/system/security/package_vuln_scan.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                    if ',' in line:
+                        package_version = max(re.findall('[0-9]+[.]+[0-9]*[.]*[0-9]', line))
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:85]:
```suggestion
                        package_version = max(re.findall('\d+\.+\d*\.*\d', line))
```

File: tests/system/security/package_vuln_scan.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    input_group.add_argument('-r', dest='input', type=str,
+                             help='specify requirements file path.')
+    input_group.add_argument('-p', dest='pip_mode', action='store_true',
+                             help='enable pip scan mode.')
+    parser.add_argument('-o', dest='output', type=str, required=True,
+                        help='specify output file.')
+    return parser.parse_args()
+
+
+def run_report(output_file_path, input_file_path):
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:27]: Shadows name output_file_path and input_file_path

File: tests/system/security/package_vuln_scan.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    json_data = {
+        'report_date': datetime.now().strftime('%d/%m/%Y - %H:%M:%S'),
+        'vulnerabilities_found': len(json_report),
+        'packages': json_report
+    }
+    with open(output_file_path, mode='w') as output_file:
+        output_file.write(json.dumps(json_data, indent=4))
+
+
+def prepare_input(pip_mode, input_file_path):
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:61]: Shadows name pip_mode and input_file_path

File: tests/system/security/package_vuln_scan.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
+import argparse
+import subprocess
+import json
+import re
+import os
+from datetime import datetime
+
+INPUT_FILE_PATH_TMP = 'input_file_tmp.txt'
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:8]: For any kind of temporary file, it is recommended to use the tempfile library.

File: tests/system/security/package_vuln_scan.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        Command line argument parsing method.
+    """
+    parser = argparse.ArgumentParser()
+    input_group = parser.add_mutually_exclusive_group(required=True)
+    input_group.add_argument('-r', dest='input', type=str,
+                             help='specify requirements file path.')
+    input_group.add_argument('-p', dest='pip_mode', action='store_true',
+                             help='enable pip scan mode.')
+    parser.add_argument('-o', dest='output', type=str, required=True,
+                        help='specify output file.')
```
Comment (AdriiiPRodri) [inline tests/system/security/package_vuln_scan.py:23]: By default, the script should provide the information through the standard output and if specified, it should be sent to a file

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]
We should investigate the possibility of using safety as a library, this way we would avoid depending on subprocess and having to hardcode the safety executable into subprocess call.

In addition, we should move the security folder and leave it next to the system folder, not inside it.

**Comment by AdriiiPRodri** [comment]
It is possible that we can use a local database, we should investigate this option.

image

**Comment by Kondent** [comment]
# UPDATE
**Pytest support added:** now this script is used by pytest in which it fails if exist a vulnerable package in the requirements.txt provided.
It also can be called with three parameters:
**--repo:** name of the repository under wazuh. Default: 'wazuh'.
**--branch:** repository branch. Default: 'master'.
**--path:** path to requirements.txt under repository. Default: 'framework/requirements.txt'


# Examples:
## Successful 
```console
⬦⬦⬦ ~/git/wazuh-qa/tests/security ⨘ python3 -m pytest --repo wazuh --branch master --path framework/requirements.txt --disable-warnings
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh-qa/tests/security
plugins: asyncio-0.15.1, html-3.1.1, metadata-1.11.0, cov-2.12.1
collected 1 item                                                                                                                                                                             

test_package_vuln_scan/test_package_vuln_scan.py .                                                                                                                                     [100%]

================================================================================ 1 passed, 1 warning in 1.03s ================================================================================
```
## Failed
```console
⬦⬦⬦ ~/git/wazuh-qa/tests/security ⨘ python3 -m pytest --repo wazuh-qa --branch master --path requirements.txt --disable-warnings
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh-qa/tests/security
plugins: asyncio-0.15.1, html-3.1.1, metadata-1.11.0, cov-2.12.1
collected 1 item                                                                                                                                                                             

test_package_vuln_scan/test_package_vuln_scan.py F                                                                                                                                     [100%]

========================================================================================== FAILURES ==========================================================================================
___________________________________________________________________________________ test_package_vuln_scan ___________________________________________________________________________________

pytestconfig = <_pytest.config.Config object at 0x7fa51a862e20>

    def test_package_vuln_scan(pytestconfig):
        branch = pytestconfig.getoption('branch')
        repo = pytestconfig.getoption('repo')
        path = pytestconfig.getoption('path')
        requirements_url = f'https://raw.githubusercontent.com/wazuh/{repo}/{branch}/{path}'
        urlretrieve(requirements_url, REQUIREMENTS_TEMP_FILE.name)
        result = report_for_pytest(REQUIREMENTS_TEMP_FILE.name)
        REQUIREMENTS_TEMP_FILE.close()
        export_report(result, REPORT_FILE)
>       assert loads(result)['vulnerabilities_found'] == 0, f'Vulnerables packages were found, full report at: ' \
                                                            f'{REPORT_FILE}'
E       AssertionError: Vulnerables packages were found, full report at: test_package_vuln_scan/report_file.json
E       assert 27 == 0

test_package_vuln_scan/test_package_vuln_scan.py:20: AssertionError
================================================================================== short test summary info ===================================================================================
FAILED test_package_vuln_scan/test_package_vuln_scan.py::test_package_vuln_scan - AssertionError: Vulnerables packages were found, full report at: test_package_vuln_scan/report_file.json
===================================================================================== 1 failed in 3.60s ======================================================================================
```
## report_file.json
<details>
  <summary>Click to expand</summary>

```json
{
    "report_date": "2021-07-28T15:25:53.503866",
    "vulnerabilities_found": 27,
    "packages": [
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.2",
            "vuln_description": "libImaging/TiffDecode.c in Pillow before 6.2.2 has a TIFF decoding integer overflow, related to realloc. See: CVE-2020-5310.",
            "safety_id": "37779"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.2",
            "vuln_description": "libImaging/FliDecode.c in Pillow before 6.2.2 has an FLI buffer overflow. See: CVE-2020-5313.",
            "safety_id": "37782"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.2",
            "vuln_description": "libImaging/SgiRleDecode.c in Pillow before 6.2.2 has an SGI buffer overflow.",
            "safety_id": "37780"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.2",
            "vuln_description": "libImaging/PcxDecode.c in Pillow before 6.2.2 has a PCX P mode buffer overflow.",
            "safety_id": "37781"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.3",
            "vuln_description": "In Pillow before 6.2.3 and 7.x before 7.0.1, there are two Buffer Overflows in libImaging/TiffDecode.c. See: CVE-2020-10379.",
            "safety_id": "38450"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<6.2.3",
            "vuln_description": "In libImaging/PcxDecode.c in Pillow before 6.2.3 and 7.x before 7.0.1, an out-of-bounds read can occur when reading PCX files where state->shuffle is instructed to read beyond state->buffer. See: CVE-2020-10378.",
            "safety_id": "38449"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<7.0.0",
            "vuln_description": "In libImaging/Jpeg2KDecode.c in Pillow before 7.0.0, there are multiple out-of-bounds reads via a crafted JP2 file. See: CVE-2020-10994.",
            "safety_id": "38451"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.0.1",
            "vuln_description": "Pillow 8.0.1 updates FreeType used in binary wheels to 2.10.4 to fix CVE-2020-15999.",
            "safety_id": "40264"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.0",
            "vuln_description": "Pillow 8.1.0 fixes TIFF OOB Write error. CVE-2020-35654 #5175.",
            "safety_id": "40265"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.0",
            "vuln_description": "Pillow 8.1.0 includes a fix for Read Overflow in PCX Decoding. CVE-2020-35653 #5174.",
            "safety_id": "40270"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.0",
            "vuln_description": "Pillow 8.1.0 includes a fix for SGI Decode buffer overrun. CVE-2020-35655 #5173.",
            "safety_id": "40271"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.1",
            "vuln_description": "Pillow 8.1.1 uses more specific regex chars to prevent ReDoS. CVE-2021-25292.",
            "safety_id": "40266"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.1",
            "vuln_description": "Pillow 8.1.1 fixes OOB Read in TiffDecode.c, and checks the tile validity before reading. CVE-2021-25291.",
            "safety_id": "40272"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.1",
            "vuln_description": "Pillow 8.1.1 fixes OOB read in SgiRleDecode.c. CVE-2021-25293.",
            "safety_id": "40273"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.1",
            "vuln_description": "Pillow 8.1.1 fixes incorrect error code checking in TiffDecode.c. CVE-2021-25289.",
            "safety_id": "40274"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.1",
            "vuln_description": "Pillow 8.1.1 fixes negative size read in TiffDecode.c. CVE-2021-25290.",
            "safety_id": "40275"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.2",
            "vuln_description": "Pillow 8.1.2 fixes memory DOS in BLP Image Plugin (CVE-2021-27921).",
            "safety_id": "40263"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.2",
            "vuln_description": "Pillow 8.1.2 fixes memory DOS in ICNS Image Plugin (CVE-2021-27922).",
            "safety_id": "40267"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.2",
            "vuln_description": "Pillow 8.1.2 fixes memory DOS in ICO Image Plugin (CVE-2021-27923).",
            "safety_id": "40268"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.1.2",
            "vuln_description": "Pillow 8.1.2 fixes memory DOS in ICNS Image Plugin (CVE-2021-27922).",
            "safety_id": "40269"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.2.0",
            "vuln_description": "An issue was discovered in Pillow before 8.2.0. There is an out-of-bounds read in J2kDecode, in j2ku_graya_la.",
            "safety_id": "40592"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.2.0",
            "vuln_description": "An issue was discovered in Pillow before 8.2.0. For BLP data, BlpImagePlugin did not properly check that reads (after jumping to file offsets) returned data. This could lead to a DoS where the decoder could be run a large number of times on empty data.",
            "safety_id": "40596"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.2.0",
            "vuln_description": "An issue was discovered in Pillow before 8.2.0. For FLI data, FliDecode did not properly check that the block advance was non-zero, potentially leading to an infinite loop on load.",
            "safety_id": "40594"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.2.0",
            "vuln_description": "An issue was discovered in Pillow before 8.2.0. There is an out-of-bounds read in J2kDecode, in j2ku_gray_i.",
            "safety_id": "40593"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<8.2.0",
            "vuln_description": "An issue was discovered in Pillow before 8.2.0. For EPS data, the readline implementation used in EPSImageFile has to deal with any combination of \\r and \\n as line endings. It used an accidentally quadratic method of accumulating lines while looking for a line ending. A malicious EPS file could use this to perform a DoS of Pillow in the open phase, before an image was accepted for opening.",
            "safety_id": "40595"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": "<=7.0.0",
            "vuln_description": "In libImaging/SgiRleDecode.c in Pillow through 7.0.0, a number of out-of-bounds reads exist in the parsing of SGI image files. See: CVE-2020-11538. Note that this is a different issue than CVE-2020-5311.",
            "safety_id": "38452"
        },
        {
            "package_name": "pillow",
            "package_version": "6.2.0",
            "package_affected_version": ">6.0,<6.2.2",
            "vuln_description": "There is a DoS vulnerability in Pillow before 6.2.2 caused by FpxImagePlugin.py calling the range function on an unvalidated 32-bit integer if the number of bands is large. On Windows running 32-bit Python, this results in an OverflowError or MemoryError due to the 2 GB limit. However, on Linux running 64-bit Python this results in the process being terminated by the OOM killer. See: CVE-2019-19911.",
            "safety_id": "37772"
        }
    ]
}
```  

</details>

File: tests/security/test_package_vuln_scan/package_vuln_scan.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+                    line = line.split(';')[0] + '\n'
+                pkg = line.strip().split('==')
+                package_list.append(package_tuple(pkg[0], pkg[1]))
+
+
+def export_report(output, output_file_path):
+    """Export report to a file or console as a message.
+
+    Args:
+        output (json): information about packages and vulnerabilities.
```
Comment (AdriiiPRodri) [inline tests/security/test_package_vuln_scan/package_vuln_scan.py:92]: This parameter is a string
```suggestion
        output (str): information about packages and vulnerabilities.
```

File: tests/security/test_package_vuln_scan/package_vuln_scan.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+    Args:
+        pip_mode (bool): enable/disable pip freeze to retrieve package information.
+        input_file_path (str): path to the input file (used if pip_mode is disabled).
+    """
+    python_process = subprocess.run([python_bin, '--version'], stdout=subprocess.PIPE, universal_newlines=True)
+    pkg = python_process.stdout.strip().split()
+    package_list.append(package_tuple(pkg[0], pkg[1]))
+    if pip_mode:
+        pip_mode_pocess = subprocess.run([python_bin, '-m', 'pip', 'freeze'], stdout=subprocess.PIPE,
```
Comment (AdriiiPRodri) [inline tests/security/test_package_vuln_scan/package_vuln_scan.py:68]:
```suggestion
        pip_mode_process = subprocess.run([python_bin, '-m', 'pip', 'freeze'], stdout=subprocess.PIPE,
```

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

**Review by AdriiiPRodri — DISMISSED** [review]
LGTM!

File: tests/security/test_package_vuln_scan/test_package_vuln_scan.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import tempfile
```
Comment (Rebits) [inline tests/security/test_package_vuln_scan/test_package_vuln_scan.py:1]: Add the Wazuh header:

```
# Copyright (C) 2015-2021, Wazuh Inc.
# Created by Wazuh, Inc. <info@wazuh.com>.
# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
```

Comment (Rebits) [reply tests/security/test_package_vuln_scan/test_package_vuln_scan.py:1]: Maybe `test_package_vuln_scan`  is not a very accurate name due to this tests check used python packages in a requirements file.  Maybe `test_python_packages_vuln_scan.py` is more descriptive.

**Review by Rebits — CHANGES_REQUESTED** [review]

**Review by Rebits — APPROVED** [review]
LGTM

**Comment by Rebits** [comment]
| Local-3.9.6 	| Status 	| Report 	|
|---	|---	|---	|
| R1 	| :green_circle:  	|  	[R1](https://github.com/wazuh/wazuh-qa/files/7114753/R1.zip) |
| R2 	| :green_circle: 	|  	[R2](https://github.com/wazuh/wazuh-qa/files/7114754/R2.zip) |
| R3 	| :green_circle: 	|  	 [R3](https://github.com/wazuh/wazuh-qa/files/7114755/R3.zip) |

**Review by Rebits — DISMISSED** [review]
LGTM

**Comment by Kondent** [comment]
# PR UPDATE
In the last commit, I did some minor fixes to adapt this test and tool to wazuh-qa repository file structure.
Overall, what I did:
* What started as a script, is a module with a couple of functions now. Available under `tools` directory and it's used by pytest.
* Few names changed (test, functions, etc).
* Read.me fixed.
Regards,
Alexis

File: tests/scans/dependencies/README.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+## How to use - Pytest
+```
+Parameters:
+    --repo: repository name. Default: 'wazuh'.
+    --branch: branch name of specified repository. Default: 'master'.
+    --requirements-path: requirements file path. Default: 'framework/requirements.txt'.
+    --report-path: output file path. Default: 'dependencies/report_file.json'.
+```
+### scanning wazuh-qa requirements file:
```
Comment (AdriiiPRodri) [inline tests/scans/dependencies/README.md:16]:
```suggestion
### Scanning wazuh-qa requirements file:
```

File: tests/scans/dependencies/README.md [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+            "safety_id": "37779"
+        },
+        ...
+        ...
+        ...
+    ]
+}
+```
+
+### scanning wazuh requirements file with a specific output path:
```
Comment (AdriiiPRodri) [inline tests/scans/dependencies/README.md:75]:
```suggestion
### Scanning wazuh requirements file with a specific output path:
```

File: tests/scans/dependencies/test_dependencies.py [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+import tempfile
+from json import loads
+from urllib.request import urlretrieve
+
+from wazuh_testing.tools.scans.dependencies import export_report, report_for_pytest
+
+REQUIREMENTS_TEMP_FILE = tempfile.NamedTemporaryFile()
+
+
+def test_python_dependencies_vuln_scan(pytestconfig):
```
Comment (AdriiiPRodri) [inline tests/scans/dependencies/test_dependencies.py:13]: We should add here a docstring with a short description for the test, something like this:

`Check that the specified dependencies do not have any known vulnerabilities`

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]
Missing docs/index.md changes

**Comment by Kondent** [comment]
> Missing docs/index.md changes

Related PR https://github.com/wazuh/wazuh-qa/pull/1659 is already handling this so I think it's no needed, let me know if I'm wrong

**Comment by Kondent** [comment]
> > Missing docs/index.md changes
> 
> Related PR #1659 is already handling this so I think it's no needed, let me know if I'm wrong

Added those changes here too. PR updated.

File: README.md [hunk 21]
```diff
@@ -19,4 +19,5 @@ In this repository you will find the tests used in the CI environment to test Wa
 - `tests`: directory containing the test suite. These are tests developed using Pytest.
     - `integration`: integration tests of the different daemons/components.
     - `system`: system tests of Wazuh.
+    - `security`: security tests used to verify Wazuh Python code and dependencies.
```
Comment (Rebits) [inline README.md:22]:
```suggestion
    - `scans`: security tests used to verify Wazuh Python code and dependencies.
```

File: tests/scans/dependencies/test_dependencies.py [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    """Check that the specified dependencies do not have any known vulnerabilities.
+
+    Args:
+        pytestconfig (fixture): Fixture that returns the :class:`_pytest.config.Config` object.
+    """
+    branch = pytestconfig.getoption('--branch')
+    repo = pytestconfig.getoption('--repo')
+    requirements_path = pytestconfig.getoption('--requirements-path')
+    report_path = pytestconfig.getoption('--report-path')
+    requirements_url = f'https://raw.githubusercontent.com/wazuh/{repo}/{branch}/{requirements_path}'
```
Comment (Rebits) [inline tests/scans/dependencies/test_dependencies.py:23]:
```suggestion
    requirements_url = f"https://raw.githubusercontent.com/wazuh/{repo}/{branch}/{requirements_path}"
```

File: tests/scans/dependencies/test_dependencies.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    urlretrieve(requirements_url, REQUIREMENTS_TEMP_FILE.name)
+    result = report_for_pytest(REQUIREMENTS_TEMP_FILE.name)
+    REQUIREMENTS_TEMP_FILE.close()
+    export_report(result, report_path)
+    assert loads(result)['vulnerabilities_found'] == 0, f'Vulnerables packages were found, full report at: ' \
+                                                        f'{report_path}'
```
Comment (Rebits) [inline tests/scans/dependencies/test_dependencies.py:29]:
```suggestion
                                                        f"{report_path}"
```

File: tests/scans/dependencies/conftest.py [hunk 24]
```diff
@@ -0,0 +1,9 @@
+import os
+
+DEFAULT_REQUIREMENTS_PATH = 'framework/requirements.txt'
+DEFAULT_REPORT_PATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'report_file.json')
+
+
+def pytest_addoption(parser):
+    parser.addoption("--requirements-path", action="store", default=DEFAULT_REQUIREMENTS_PATH)
+    parser.addoption("--report-path", action="store", default=DEFAULT_REPORT_PATH)
```
Comment (Rebits) [inline tests/scans/dependencies/conftest.py:9]:
```suggestion
    parser.addoption('--requirements-path', action='store', default=DEFAULT_REQUIREMENTS_PATH)
    parser.addoption('--report-path', action='store', default=DEFAULT_REPORT_PATH)
```

File: tests/scans/conftest.py [hunk 25]
```diff
@@ -0,0 +1,7 @@
+DEFAULT_BRANCH = 'master'
+DEFAULT_REPOSITORY = 'wazuh'
+
+
+def pytest_addoption(parser):
+    parser.addoption("--branch", action="store", default=DEFAULT_BRANCH)
+    parser.addoption("--repo", action="store", default=DEFAULT_REPOSITORY)
```
Comment (Rebits) [inline tests/scans/conftest.py:7]:
```suggestion
    parser.addoption('--branch', action='store', default=DEFAULT_BRANCH)
    parser.addoption('--repo', action='store', default=DEFAULT_REPOSITORY)
```

**Review by Rebits — CHANGES_REQUESTED** [review]
Minor changes are required

**Comment by Rebits** [comment]
| Local-3.9.6 	| Status 	| 
|---	|---	|
| R1 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7170523/r1.zip) 	| 
| R2 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7170525/r2.zip) 	|  	
| R3 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7170526/r3.zip) 	|

**Review by AdriiiPRodri — DISMISSED** [review]
LGTM

**Review by davidjiglesias — APPROVED** [review]

**Comment by Rebits** [comment]
| Local-3.9.6 	| Status 	| 
|---	|---	|
| R1 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7249795/R1.zip)| 
| R2 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7249796/R2.zip)	|  	
| R3 	| [:green_circle:](https://github.com/wazuh/wazuh-qa/files/7249797/R3.zip)|

**Review by Rebits — APPROVED** [review]
LGTM

**Pull request merged**